// repomix/lib/shared/errorHandle.js
import { inspect } from 'node:util';
import { z } from 'zod';
import { REPOMIX_DISCORD_URL, REPOMIX_ISSUES_URL } from './constants.js';
import { logger, repomixLogLevels } from './logger.js';
export class RepomixError extends Error {
    constructor(message, options) {
        super(message, options);
        this.name = 'RepomixError';
    }
}
export class RepomixConfigValidationError extends RepomixError {
    constructor(message, options) {
        super(message, options);
        this.name = 'RepomixConfigValidationError';
    }
}
export class OperationCancelledError extends RepomixError {
    constructor(message = 'Operation cancelled') {
        super(message);
        this.name = 'OperationCancelledError';
    }
}
export const handleError = (error) => {
    logger.log('');
    if (isRepomixError(error)) {
        logger.error(`✖ ${error.message}`);
        if (logger.getLogLevel() < repomixLogLevels.DEBUG) {
            logger.log('');
            logger.note('For detailed debug information, use the --verbose flag');
        }
        logger.debug('Stack trace:', error.stack);
        if (error.cause) {
            logger.debug('Caused by:', error.cause);
        }
    }
    else if (isError(error)) {
        logger.error(`✖ Unexpected error: ${error.message}`);
        logger.note('Stack trace:', error.stack);
        if (logger.getLogLevel() < repomixLogLevels.DEBUG) {
            logger.log('');
            logger.note('For detailed debug information, use the --verbose flag');
        }
    }
    else {
        logger.error('✖ An unknown error occurred');
        try {
            logger.note('Error details:', inspect(error, {
                depth: 3,
                colors: false,
                maxArrayLength: 10,
                maxStringLength: 200,
                breakLength: Number.POSITIVE_INFINITY,
            }));
        }
        catch {
            logger.note('Error details: [Error object could not be serialized]');
        }
        if (logger.getLogLevel() < repomixLogLevels.DEBUG) {
            logger.log('');
            logger.note('For detailed debug information, use the --verbose flag');
        }
    }
    logger.log('');
    logger.info('Need help?');
    logger.info(`• File an issue on GitHub: ${REPOMIX_ISSUES_URL}`);
    logger.info(`• Join our Discord community: ${REPOMIX_DISCORD_URL}`);
};
const isError = (error) => {
    if (error instanceof Error)
        return true;
    if (typeof error !== 'object' || error === null)
        return false;
    const obj = error;
    return (typeof obj.message === 'string' &&
        (!('stack' in obj) || typeof obj.stack === 'string') &&
        (!('name' in obj) || typeof obj.name === 'string'));
};
const isRepomixError = (error) => {
    if (error instanceof RepomixError)
        return true;
    if (typeof error !== 'object' || error === null)
        return false;
    const obj = error;
    return (typeof obj.message === 'string' &&
        'name' in obj &&
        (obj.name === RepomixError.name || obj.name === RepomixConfigValidationError.name));
};
export const rethrowValidationErrorIfZodError = (error, message) => {
    if (error instanceof z.ZodError) {
        const zodErrorText = error.issues.map((err) => `[${err.path.join('.')}] ${err.message}`).join('\n  ');
        throw new RepomixConfigValidationError(`${message}\n\n  ${zodErrorText}\n\n  Please check the config file and try again.`);
    }
};
