// repomix/lib/core/metrics/calculateMetrics.js
import { initTaskRunner } from '../../shared/processConcurrency.js';
import { buildSplitOutputFilePath } from '../output/outputSplit.js';
import { calculateGitDiffMetrics } from './calculateGitDiffMetrics.js';
import { calculateGitLogMetrics } from './calculateGitLogMetrics.js';
import { calculateOutputMetrics } from './calculateOutputMetrics.js';
import { calculateSelectiveFileMetrics } from './calculateSelectiveFileMetrics.js';
export const calculateMetrics = async (processedFiles, output, progressCallback, config, gitDiffResult, gitLogResult, deps = {
    calculateSelectiveFileMetrics,
    calculateOutputMetrics,
    calculateGitDiffMetrics,
    calculateGitLogMetrics,
    taskRunner: undefined,
}) => {
    progressCallback('Calculating metrics...');
    const taskRunner = deps.taskRunner ??
        initTaskRunner({
            numOfTasks: processedFiles.length,
            workerType: 'calculateMetrics',
            runtime: 'worker_threads',
        });
    try {
        const outputParts = Array.isArray(output) ? output : [output];
        const topFilesLength = config.output.topFilesLength;
        const shouldCalculateAllFiles = !!config.output.tokenCountTree;
        const metricsTargetPaths = shouldCalculateAllFiles
            ? processedFiles.map((file) => file.path)
            : [...processedFiles]
                .sort((a, b) => b.content.length - a.content.length)
                .slice(0, Math.min(processedFiles.length, Math.max(topFilesLength * 10, topFilesLength)))
                .map((file) => file.path);
        const [selectiveFileMetrics, outputTokenCounts, gitDiffTokenCount, gitLogTokenCount] = await Promise.all([
            deps.calculateSelectiveFileMetrics(processedFiles, metricsTargetPaths, config.tokenCount.encoding, progressCallback, { taskRunner }),
            Promise.all(outputParts.map(async (part, index) => {
                const partPath = outputParts.length > 1
                    ? buildSplitOutputFilePath(config.output.filePath, index + 1)
                    : config.output.filePath;
                return await deps.calculateOutputMetrics(part, config.tokenCount.encoding, partPath, { taskRunner });
            })),
            deps.calculateGitDiffMetrics(config, gitDiffResult, { taskRunner }),
            deps.calculateGitLogMetrics(config, gitLogResult, { taskRunner }),
        ]);
        const totalTokens = outputTokenCounts.reduce((sum, count) => sum + count, 0);
        const totalFiles = processedFiles.length;
        const totalCharacters = outputParts.reduce((sum, part) => sum + part.length, 0);
        const fileCharCounts = {};
        for (const file of processedFiles) {
            fileCharCounts[file.path] = file.content.length;
        }
        const fileTokenCounts = {};
        for (const file of selectiveFileMetrics) {
            fileTokenCounts[file.path] = file.tokenCount;
        }
        return {
            totalFiles,
            totalCharacters,
            totalTokens,
            fileCharCounts,
            fileTokenCounts,
            gitDiffTokenCount: gitDiffTokenCount,
            gitLogTokenCount: gitLogTokenCount.gitLogTokenCount,
        };
    }
    finally {
        if (!deps.taskRunner) {
            await taskRunner.cleanup();
        }
    }
};
