// repomix/lib/cli/cliSpinner.js
import logUpdate from 'log-update';
import pc from 'picocolors';
const dotsFrames = ['‚†č', '‚†ô', '‚†Ļ', '‚†ł', '‚†ľ', '‚†ī', '‚†¶', '‚†ß', '‚†á', '‚†Ź'];
const dotsInterval = 80;
export class Spinner {
    message;
    currentFrame = 0;
    interval = null;
    isQuiet;
    constructor(message, cliOptions) {
        this.message = message;
        this.isQuiet = cliOptions?.quiet || cliOptions?.verbose || cliOptions?.stdout || false;
    }
    start() {
        if (this.isQuiet) {
            return;
        }
        const framesLength = dotsFrames.length;
        this.interval = setInterval(() => {
            this.currentFrame++;
            const frame = dotsFrames[this.currentFrame % framesLength];
            logUpdate(`${pc.cyan(frame)} ${this.message}`);
        }, dotsInterval);
    }
    update(message) {
        if (this.isQuiet) {
            return;
        }
        this.message = message;
    }
    stop(finalMessage) {
        if (this.isQuiet) {
            return;
        }
        if (this.interval) {
            clearInterval(this.interval);
            this.interval = null;
        }
        logUpdate(finalMessage);
        logUpdate.done();
    }
    succeed(message) {
        if (this.isQuiet) {
            return;
        }
        this.stop(`${pc.green('‚úĒ')} ${message}`);
    }
    fail(message) {
        if (this.isQuiet) {
            return;
        }
        this.stop(`${pc.red('‚úĖ')} ${message}`);
    }
}
