// repomix/lib/core/security/securityCheck.js
import pc from 'picocolors';
import { logger } from '../../shared/logger.js';
import { initTaskRunner } from '../../shared/processConcurrency.js';
export const runSecurityCheck = async (rawFiles, progressCallback = () => { }, gitDiffResult, gitLogResult, deps = {
    initTaskRunner,
}) => {
    const gitDiffTasks = [];
    const gitLogTasks = [];
    if (gitDiffResult) {
        if (gitDiffResult.workTreeDiffContent) {
            gitDiffTasks.push({
                filePath: 'Working tree changes',
                content: gitDiffResult.workTreeDiffContent,
                type: 'gitDiff',
            });
        }
        if (gitDiffResult.stagedDiffContent) {
            gitDiffTasks.push({
                filePath: 'Staged changes',
                content: gitDiffResult.stagedDiffContent,
                type: 'gitDiff',
            });
        }
    }
    if (gitLogResult) {
        if (gitLogResult.logContent) {
            gitLogTasks.push({
                filePath: 'Git log history',
                content: gitLogResult.logContent,
                type: 'gitLog',
            });
        }
    }
    const taskRunner = deps.initTaskRunner({
        numOfTasks: rawFiles.length + gitDiffTasks.length + gitLogTasks.length,
        workerType: 'securityCheck',
        runtime: 'worker_threads',
    });
    const fileTasks = rawFiles.map((file) => ({
        filePath: file.path,
        content: file.content,
        type: 'file',
    }));
    const tasks = [...fileTasks, ...gitDiffTasks, ...gitLogTasks];
    try {
        logger.trace(`Starting security check for ${tasks.length} files/content`);
        const startTime = process.hrtime.bigint();
        let completedTasks = 0;
        const totalTasks = tasks.length;
        const results = await Promise.all(tasks.map((task) => taskRunner.run(task).then((result) => {
            completedTasks++;
            progressCallback(`Running security check... (${completedTasks}/${totalTasks}) ${pc.dim(task.filePath)}`);
            logger.trace(`Running security check... (${completedTasks}/${totalTasks}) ${task.filePath}`);
            return result;
        })));
        const endTime = process.hrtime.bigint();
        const duration = Number(endTime - startTime) / 1e6;
        logger.trace(`Security check completed in ${duration.toFixed(2)}ms`);
        return results.filter((result) => result !== null);
    }
    catch (error) {
        logger.error('Error during security check:', error);
        throw error;
    }
    finally {
        await taskRunner.cleanup();
    }
};
