// repomix/lib/core/security/workers/securityCheckWorker.js
import { lintSource } from '@secretlint/core';
import { creator } from '@secretlint/secretlint-rule-preset-recommend';
import { logger, setLogLevelByWorkerData } from '../../../shared/logger.js';
setLogLevelByWorkerData();
export default async ({ filePath, content, type }) => {
    const config = createSecretLintConfig();
    try {
        const processStartAt = process.hrtime.bigint();
        const secretLintResult = await runSecretLint(filePath, content, type, config);
        const processEndAt = process.hrtime.bigint();
        logger.trace(`Checked security on ${filePath}. Took: ${(Number(processEndAt - processStartAt) / 1e6).toFixed(2)}ms`);
        return secretLintResult;
    }
    catch (error) {
        logger.error(`Error checking security on ${filePath}:`, error);
        throw error;
    }
};
export const runSecretLint = async (filePath, content, type, config) => {
    const result = await lintSource({
        source: {
            filePath: filePath,
            content: content,
            ext: filePath.split('.').pop() || '',
            contentType: 'text',
        },
        options: {
            config: config,
        },
    });
    if (result.messages.length > 0) {
        const issueCount = result.messages.length;
        const issueText = issueCount === 1 ? 'security issue' : 'security issues';
        logger.trace(`Found ${issueCount} ${issueText} in ${filePath}`);
        return {
            filePath,
            messages: result.messages.map((message) => message.message),
            type,
        };
    }
    return null;
};
export const createSecretLintConfig = () => ({
    rules: [
        {
            id: '@secretlint/secretlint-rule-preset-recommend',
            rule: creator,
        },
    ],
});
export const onWorkerTermination = async () => {
};
